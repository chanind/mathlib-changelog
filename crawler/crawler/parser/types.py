from __future__ import annotations
from dataclasses import dataclass
from typing import Literal


ItemType = Literal["theorem", "def", "inductive", "structure"]
BlockType = Literal["section", "namespace"]


@dataclass
class ParsedItem:
    type: ItemType
    name: str
    namespace: list[str]
    line_hash: str

    @property
    def full_name(self) -> str:
        return self.type + ":" + ".".join([*self.namespace, self.name])


@dataclass
class LeanParseResult:
    items: list[ParsedItem]

    def get_item(self, full_name: str) -> ParsedItem | None:
        for item in self.items:
            if item.full_name == full_name:
                return item
        return None


@dataclass
class ParseState:
    blocks: list[tuple[BlockType, str]]
    items: list[ParsedItem]

    @property
    def namespace(self) -> list[str]:
        namespace = []
        for block in self.blocks:
            if block[0] == "namespace":
                # handle cases like "namespace Thing1.thing2"
                namespace += block[1].split(".")
        return namespace
