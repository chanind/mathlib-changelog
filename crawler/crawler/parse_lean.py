from __future__ import annotations
from dataclasses import dataclass
import re
from typing import Literal, cast
import hashlib


ItemType = Literal["lemma", "theorem", "def"]
BlockType = Literal["section", "namespace"]


@dataclass
class ParsedItem:
    type: ItemType
    name: str
    namespace: list[str]
    line_hash: str

    @property
    def full_name(self) -> str:
        return self.type + ":" + ".".join([*self.namespace, self.name])


@dataclass
class LeanParseResult:
    items: list[ParsedItem]

    def get_item(self, full_name: str) -> ParsedItem | None:
        for item in self.items:
            if item.full_name == full_name:
                return item
        return None


@dataclass
class ParseState:
    blocks: list[tuple[BlockType, str]]
    items: list[ParsedItem]

    @property
    def namespace(self) -> list[str]:
        namespace = []
        for block in self.blocks:
            if block[0] == "namespace":
                # handle cases like "namespace Thing1.thing2"
                namespace += block[1].split(".")
        return namespace


def parse_lean(contents: str) -> list[ParsedItem]:
    state = ParseState(blocks=[], items=[])
    lines = strip_lean_comments(contents).splitlines()
    for line in lines:
        parse_line(line, state)
    if len(state.blocks) > 0:
        raise Exception("Invalid parse state! open blocks remaining at end of parse")
    return state.items


OPTIONAL_DECORATOR_REGEX = r"(?:@\[[^\]]+\]\s?)"
ITEM_REGEX = rf"^\s*{OPTIONAL_DECORATOR_REGEX}*(def|lemma|theorem)\s+([^\s]+)"
START_BLOCK_REGEX = r"^\s*(section|namespace)\s+([^\s]+)"
END_BLOCK_REGEX = r"^\s*end\s+([^\s]+)"


def parse_line(line: str, state: ParseState) -> None:
    """
    Extract any items from the line, and track sections/namespaces.
    This will modify the state in-place.

    NOTE: this assumes there can only be a single def/theorem/lemma/section/namespace per line
    """
    item_match = re.search(ITEM_REGEX, line)
    start_block_match = re.search(START_BLOCK_REGEX, line)
    end_block_match = re.search(END_BLOCK_REGEX, line)

    if item_match:
        line_hash = hashlib.md5(line.encode()).hexdigest()
        [item_type, item_name] = item_match.groups()
        state.items.append(
            build_parsed_item(item_type, item_name, state.namespace, line_hash)
        )
    if start_block_match:
        [block_type, block_name] = start_block_match.groups()
        state.blocks.append((cast(BlockType, block_type), block_name))
    if end_block_match:
        [block_name] = end_block_match.groups()
        if block_name != state.blocks[-1][1]:
            raise Exception(f"Tried to pop unopened block {block_name}")
        state.blocks.pop()


def build_parsed_item(
    item_type: str,
    item_name: str,
    namespace: list[str],
    line_hash: str,
) -> ParsedItem:
    item_name_parts = item_name.split(".")

    if item_type not in ["lemma", "theorem", "def"]:
        raise Exception(f"Invalid item type found: {item_type}")
    safe_item_type = cast(ItemType, item_type)

    # starting with _root_ means we should throw out the namespace we're in
    if item_name_parts[0] == "_root_":
        return ParsedItem(
            type=safe_item_type,
            name=item_name_parts[-1],
            namespace=item_name_parts[1:-1],
            line_hash=line_hash,
        )
    return ParsedItem(
        type=safe_item_type,
        name=item_name_parts[-1],
        namespace=namespace + item_name_parts[0:-1],
        line_hash=line_hash,
    )


def strip_lean_comments(lean_contents: str) -> str:
    processed = re.sub(r"/-((?!-/)(.|\r|\n))*-/", "", lean_contents, re.MULTILINE)
    return re.sub(r"\-\-[^\n\r]*", "", processed, re.MULTILINE)
