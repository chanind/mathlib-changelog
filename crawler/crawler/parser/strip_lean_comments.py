import re


def strip_lean_comments(lean_contents: str) -> str:
    processed = re.sub(r"/-((?!-/)(.|\r|\n))*-/", "", lean_contents, re.MULTILINE)
    return re.sub(r"\-\-[^\n\r]*", "", processed, re.MULTILINE)
