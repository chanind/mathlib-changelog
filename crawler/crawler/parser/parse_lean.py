from __future__ import annotations
from dataclasses import dataclass
import logging
import re
from typing import Literal, cast
import hashlib

from .errors import LeanParseError
from .strip_lean_comments import strip_lean_comments


ItemType = Literal["lemma", "theorem", "def"]
BlockType = Literal["section", "namespace"]


@dataclass
class ParsedItem:
    type: ItemType
    name: str
    namespace: list[str]
    line_hash: str

    @property
    def full_name(self) -> str:
        return self.type + ":" + ".".join([*self.namespace, self.name])


@dataclass
class LeanParseResult:
    items: list[ParsedItem]

    def get_item(self, full_name: str) -> ParsedItem | None:
        for item in self.items:
            if item.full_name == full_name:
                return item
        return None


@dataclass
class ParseState:
    blocks: list[tuple[BlockType, str]]
    items: list[ParsedItem]

    @property
    def namespace(self) -> list[str]:
        namespace = []
        for block in self.blocks:
            if block[0] == "namespace":
                # handle cases like "namespace Thing1.thing2"
                namespace += block[1].split(".")
        return namespace


def parse_lean(contents: str) -> list[ParsedItem]:
    state = ParseState(blocks=[], items=[])

    lines = strip_lean_comments(contents).splitlines()
    for line in lines:
        parse_line(line, state)
    if len(state.namespace) > 0:
        logging.warning(
            f"Invalid parse state! open namespace remaining at end of parse: {state.namespace}"
        )
    return state.items


OPTIONAL_DECORATOR_REGEX = r"(?:@\[[^\]]+\]\s?)"
ITEM_REGEX = rf"^\s*{OPTIONAL_DECORATOR_REGEX}*(def|lemma|theorem)\s+([^\s]+)"
START_BLOCK_REGEX = r"^\s*(section|namespace)\s+([^\s]+)"
END_BLOCK_REGEX = r"^\s*end\s+([^\s]+)"


def parse_line(line: str, state: ParseState) -> None:
    """
    Extract any items from the line, and track sections/namespaces.
    This will modify the state in-place.

    NOTE: this assumes there can only be a single def/theorem/lemma/section/namespace per line
    """
    item_match = re.search(ITEM_REGEX, line)
    start_block_match = re.search(START_BLOCK_REGEX, line)
    end_block_match = re.search(END_BLOCK_REGEX, line)

    if item_match:
        line_hash = hashlib.md5(line.encode()).hexdigest()
        [item_type, item_name] = item_match.groups()
        state.items.append(
            build_parsed_item(item_type, item_name, state.namespace, line_hash)
        )
    if start_block_match:
        [block_type, block_name] = start_block_match.groups()
        # weird keyword `section end` in https://github.com/leanprover-community/mathlib/blob/master/src/algebraic_geometry/pullbacks.lean#L113
        # not sure what this means, so just skip it
        if block_type == "section" and block_name == "end":
            return
        state.blocks.append((cast(BlockType, block_type), block_name))
    if end_block_match:
        [block_name] = end_block_match.groups()
        if len(state.blocks) > 0 and block_name == state.blocks[-1][1]:
            state.blocks.pop()


def build_parsed_item(
    item_type: str,
    item_name: str,
    namespace: list[str],
    line_hash: str,
) -> ParsedItem:
    item_name_parts = item_name.split(".")

    if item_type not in ["lemma", "theorem", "def"]:
        raise Exception(f"Invalid item type found: {item_type}")
    safe_item_type = cast(ItemType, item_type)

    # starting with _root_ means we should throw out the namespace we're in
    if item_name_parts[0] == "_root_":
        return ParsedItem(
            type=safe_item_type,
            name=item_name_parts[-1],
            namespace=item_name_parts[1:-1],
            line_hash=line_hash,
        )
    return ParsedItem(
        type=safe_item_type,
        name=item_name_parts[-1],
        namespace=namespace + item_name_parts[0:-1],
        line_hash=line_hash,
    )
