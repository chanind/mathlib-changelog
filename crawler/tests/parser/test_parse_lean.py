from pathlib import Path
from crawler.parser.parse_lean import parse_lean


FIXTURES_DIR = Path(__file__).resolve().parent / "../fixtures"


def read_fixture(filename: str) -> str:
    with open(FIXTURES_DIR / filename, "r") as file:
        return file.read()


def test_parse_lean_ess_sup_fixture() -> None:
    contents = read_fixture("ess_sup.lean")
    parsed_items = parse_lean(contents)
    assert [item.full_name for item in parsed_items] == [
        "def:ess_sup",
        "def:ess_inf",
        "lemma:ess_sup_congr_ae",
        "lemma:ess_inf_congr_ae",
        "lemma:ess_sup_eq_Inf",
        "lemma:ess_sup_measure_zero",
        "lemma:ess_inf_measure_zero",
        "lemma:ess_sup_mono_ae",
        "lemma:ess_inf_mono_ae",
        "lemma:ess_sup_const",
        "lemma:ess_sup_le_of_ae_le",
        "lemma:ess_inf_const",
        "lemma:le_ess_inf_of_ae_le",
        "lemma:ess_sup_const_bot",
        "lemma:ess_inf_const_top",
        "lemma:order_iso.ess_sup_apply",
        "lemma:order_iso.ess_inf_apply",
        "lemma:ess_sup_mono_measure",
        "lemma:ess_sup_mono_measure'",
        "lemma:ess_inf_antitone_measure",
        "lemma:ess_sup_smul_measure",
        "lemma:ess_sup_comp_le_ess_sup_map_measure",
        "lemma:measurable_embedding.ess_sup_map_measure",
        "lemma:ess_sup_map_measure_of_measurable",
        "lemma:ess_sup_map_measure",
        "lemma:ae_lt_of_ess_sup_lt",
        "lemma:ae_lt_of_lt_ess_inf",
        "lemma:ess_sup_indicator_eq_ess_sup_restrict",
        "lemma:ennreal.ae_le_ess_sup",
        "lemma:ennreal.ess_sup_eq_zero_iff",
        "lemma:ennreal.ess_sup_const_mul",
        "lemma:ennreal.ess_sup_add_le",
        "lemma:ennreal.ess_sup_liminf_le",
    ]


def test_parse_lean_filtered_colimits_fixture() -> None:
    contents = read_fixture("filtered_colimits.lean")
    parsed_items = parse_lean(contents)
    assert [item.full_name for item in parsed_items] == [
        "def:SemiRing.filtered_colimits.colimit",
        "def:SemiRing.filtered_colimits.colimit_cocone",
        "def:SemiRing.filtered_colimits.colimit_cocone_is_colimit",
        "def:CommSemiRing.filtered_colimits.colimit",
        "def:CommSemiRing.filtered_colimits.colimit_cocone",
        "def:CommSemiRing.filtered_colimits.colimit_cocone_is_colimit",
        "def:Ring.filtered_colimits.colimit",
        "def:Ring.filtered_colimits.colimit_cocone",
        "def:Ring.filtered_colimits.colimit_cocone_is_colimit",
        "def:CommRing.filtered_colimits.colimit",
        "def:CommRing.filtered_colimits.colimit_cocone",
        "def:CommRing.filtered_colimits.colimit_cocone_is_colimit",
    ]


def test_parse_lean_algebra_fixture() -> None:
    contents = read_fixture("algebra.lean")
    parsed_items = parse_lean(contents)
    assert [item.full_name for item in parsed_items] == [
        "def:category_theory.monad.algebra.hom.id",
        "def:category_theory.monad.algebra.hom.comp",
        "lemma:category_theory.monad.algebra.comp_eq_comp",
        "lemma:category_theory.monad.algebra.id_eq_id",
        "lemma:category_theory.monad.algebra.id_f",
        "lemma:category_theory.monad.algebra.comp_f",
        "def:category_theory.monad.algebra.iso_mk",
        "def:category_theory.monad.forget",
        "def:category_theory.monad.free",
        "def:category_theory.monad.adj",
        "lemma:category_theory.monad.algebra_iso_of_iso",
        "lemma:category_theory.monad.algebra_epi_of_epi",
        "lemma:category_theory.monad.algebra_mono_of_mono",
        "lemma:category_theory.monad.left_adjoint_forget",
        "lemma:category_theory.monad.of_right_adjoint_forget",
        "def:category_theory.monad.algebra_functor_of_monad_hom",
        "def:category_theory.monad.algebra_functor_of_monad_hom_id",
        "def:category_theory.monad.algebra_functor_of_monad_hom_comp",
        "def:category_theory.monad.algebra_functor_of_monad_hom_eq",
        "def:category_theory.monad.algebra_equiv_of_iso_monads",
        "lemma:category_theory.monad.algebra_equiv_of_iso_monads_comp_forget",
        "def:category_theory.comonad.coalgebra.hom.id",
        "def:category_theory.comonad.coalgebra.hom.comp",
        "lemma:category_theory.comonad.coalgebra.comp_eq_comp",
        "lemma:category_theory.comonad.coalgebra.id_eq_id",
        "lemma:category_theory.comonad.coalgebra.id_f",
        "lemma:category_theory.comonad.coalgebra.comp_f",
        "def:category_theory.comonad.coalgebra.iso_mk",
        "def:category_theory.comonad.forget",
        "def:category_theory.comonad.cofree",
        "def:category_theory.comonad.adj",
        "lemma:category_theory.comonad.coalgebra_iso_of_iso",
        "lemma:category_theory.comonad.algebra_epi_of_epi",
        "lemma:category_theory.comonad.algebra_mono_of_mono",
        "lemma:category_theory.comonad.right_adjoint_forget",
        "lemma:category_theory.comonad.of_left_adjoint_forget",
    ]


def test_parse_lean_cont_mdiff_fixture() -> None:
    contents = read_fixture("cont_mdiff.lean")
    parse_lean(contents)  # just make sure it doesn't error


def test_parse_lean_to_additive_fixture() -> None:
    contents = read_fixture("to_additive.lean")
    parse_lean(contents)  # just make sure it doesn't error


def test_parse_lean_fermat4_fixture() -> None:
    contents = read_fixture("fermat4.lean")
    parse_lean(contents)  # just make sure it doesn't error
