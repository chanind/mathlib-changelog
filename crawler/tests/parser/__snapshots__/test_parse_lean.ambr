# name: test_parse_lean_algebra_fixture
  list([
    'structure:category_theory.monad.algebra',
    'structure:category_theory.monad.algebra.hom',
    'def:category_theory.monad.algebra.hom.id',
    'def:category_theory.monad.algebra.hom.comp',
    'lemma:category_theory.monad.algebra.comp_eq_comp',
    'lemma:category_theory.monad.algebra.id_eq_id',
    'lemma:category_theory.monad.algebra.id_f',
    'lemma:category_theory.monad.algebra.comp_f',
    'def:category_theory.monad.algebra.iso_mk',
    'def:category_theory.monad.forget',
    'def:category_theory.monad.free',
    'def:category_theory.monad.adj',
    'lemma:category_theory.monad.algebra_iso_of_iso',
    'lemma:category_theory.monad.algebra_epi_of_epi',
    'lemma:category_theory.monad.algebra_mono_of_mono',
    'lemma:category_theory.monad.left_adjoint_forget',
    'lemma:category_theory.monad.of_right_adjoint_forget',
    'def:category_theory.monad.algebra_functor_of_monad_hom',
    'def:category_theory.monad.algebra_functor_of_monad_hom_id',
    'def:category_theory.monad.algebra_functor_of_monad_hom_comp',
    'def:category_theory.monad.algebra_functor_of_monad_hom_eq',
    'def:category_theory.monad.algebra_equiv_of_iso_monads',
    'lemma:category_theory.monad.algebra_equiv_of_iso_monads_comp_forget',
    'structure:category_theory.comonad.coalgebra',
    'structure:category_theory.comonad.coalgebra.hom',
    'def:category_theory.comonad.coalgebra.hom.id',
    'def:category_theory.comonad.coalgebra.hom.comp',
    'lemma:category_theory.comonad.coalgebra.comp_eq_comp',
    'lemma:category_theory.comonad.coalgebra.id_eq_id',
    'lemma:category_theory.comonad.coalgebra.id_f',
    'lemma:category_theory.comonad.coalgebra.comp_f',
    'def:category_theory.comonad.coalgebra.iso_mk',
    'def:category_theory.comonad.forget',
    'def:category_theory.comonad.cofree',
    'def:category_theory.comonad.adj',
    'lemma:category_theory.comonad.coalgebra_iso_of_iso',
    'lemma:category_theory.comonad.algebra_epi_of_epi',
    'lemma:category_theory.comonad.algebra_mono_of_mono',
    'lemma:category_theory.comonad.right_adjoint_forget',
    'lemma:category_theory.comonad.of_left_adjoint_forget',
  ])
# ---
# name: test_parse_lean_cont_mdiff_fixture
  list([
    'def:cont_diff_within_at_prop',
    'lemma:cont_diff_within_at_local_invariant_prop',
    'lemma:cont_diff_within_at_local_invariant_prop_mono',
    'lemma:cont_diff_within_at_local_invariant_prop_id',
    'def:cont_mdiff_within_at',
    'def:smooth_within_at',
    'def:cont_mdiff_at',
    'def:smooth_at',
    'def:cont_mdiff_on',
    'def:smooth_on',
    'def:cont_mdiff',
    'def:smooth',
    'lemma:cont_mdiff.smooth',
    'lemma:smooth.cont_mdiff',
    'lemma:cont_mdiff_on.smooth_on',
    'lemma:smooth_on.cont_mdiff_on',
    'lemma:cont_mdiff_at.smooth_at',
    'lemma:smooth_at.cont_mdiff_at',
    'lemma:cont_mdiff_within_at.smooth_within_at',
    'lemma:smooth_within_at.cont_mdiff_within_at',
    'lemma:cont_mdiff.cont_mdiff_at',
    'lemma:smooth.smooth_at',
    'lemma:cont_mdiff_within_at_univ',
    'lemma:smooth_at_univ',
    'lemma:cont_mdiff_on_univ',
    'lemma:smooth_on_univ',
    'lemma:cont_mdiff_within_at_iff',
    "lemma:cont_mdiff_within_at_iff''",
    'lemma:cont_mdiff_within_at_iff_target',
    'lemma:smooth_within_at_iff',
    'lemma:smooth_within_at_iff_target',
    'lemma:cont_mdiff_at_ext_chart_at',
    "lemma:cont_mdiff_within_at_iff'",
    "lemma:cont_mdiff_at_ext_chart_at'",
    'lemma:cont_mdiff_on_iff',
    'lemma:cont_mdiff_on_iff_target',
    'lemma:smooth_on_iff',
    'lemma:smooth_on_iff_target',
    'lemma:cont_mdiff_iff',
    'lemma:cont_mdiff_iff_target',
    'lemma:smooth_iff',
    'lemma:smooth_iff_target',
    'lemma:cont_mdiff_within_at.of_le',
    'lemma:cont_mdiff_at.of_le',
    'lemma:cont_mdiff_on.of_le',
    'lemma:cont_mdiff.of_le',
    'lemma:cont_mdiff_within_at.of_succ',
    'lemma:cont_mdiff_at.of_succ',
    'lemma:cont_mdiff_on.of_succ',
    'lemma:cont_mdiff.of_succ',
    'lemma:cont_mdiff_within_at.continuous_within_at',
    'lemma:cont_mdiff_at.continuous_at',
    'lemma:cont_mdiff_on.continuous_on',
    'lemma:cont_mdiff.continuous',
    'lemma:cont_mdiff_within_at.mdifferentiable_within_at',
    'lemma:cont_mdiff_at.mdifferentiable_at',
    'lemma:cont_mdiff_on.mdifferentiable_on',
    'lemma:cont_mdiff.mdifferentiable',
    'lemma:smooth.mdifferentiable',
    'lemma:smooth.mdifferentiable_at',
    'lemma:smooth.mdifferentiable_within_at',
    'lemma:cont_mdiff_within_at_top',
    'lemma:cont_mdiff_at_top',
    'lemma:cont_mdiff_on_top',
    'lemma:cont_mdiff_top',
    'lemma:cont_mdiff_within_at_iff_nat',
    'lemma:cont_mdiff_within_at.mono',
    'lemma:cont_mdiff_at.cont_mdiff_within_at',
    'lemma:smooth_at.smooth_within_at',
    'lemma:cont_mdiff_on.mono',
    'lemma:cont_mdiff.cont_mdiff_on',
    'lemma:smooth.smooth_on',
    "lemma:cont_mdiff_within_at_inter'",
    'lemma:cont_mdiff_within_at_inter',
    'lemma:cont_mdiff_within_at.cont_mdiff_at',
    'lemma:smooth_within_at.smooth_at',
    'lemma:cont_mdiff_on_ext_chart_at',
    'lemma:cont_mdiff_within_at_iff_cont_mdiff_on_nhds',
    'lemma:cont_mdiff_at_iff_cont_mdiff_on_nhds',
    'lemma:cont_mdiff_within_at.congr',
    'lemma:cont_mdiff_within_at_congr',
    'lemma:cont_mdiff_within_at.congr_of_eventually_eq',
    'lemma:filter.eventually_eq.cont_mdiff_within_at_iff',
    'lemma:cont_mdiff_at.congr_of_eventually_eq',
    'lemma:filter.eventually_eq.cont_mdiff_at_iff',
    'lemma:cont_mdiff_on.congr',
    'lemma:cont_mdiff_on_congr',
    'lemma:cont_mdiff_on_of_locally_cont_mdiff_on',
    'lemma:cont_mdiff_of_locally_cont_mdiff_on',
    'lemma:cont_mdiff_within_at.comp',
    'lemma:cont_mdiff_on.comp',
    "lemma:cont_mdiff_on.comp'",
    'lemma:cont_mdiff.comp',
    "lemma:cont_mdiff_within_at.comp'",
    'lemma:cont_mdiff_at.comp_cont_mdiff_within_at',
    'lemma:cont_mdiff_at.comp',
    'lemma:cont_mdiff.comp_cont_mdiff_on',
    'lemma:smooth.comp_smooth_on',
    'lemma:cont_mdiff_on.comp_cont_mdiff',
    'lemma:smooth_on.comp_smooth',
    'lemma:cont_mdiff_on_of_mem_maximal_atlas',
    'lemma:cont_mdiff_on_symm_of_mem_maximal_atlas',
    'lemma:cont_mdiff_on_chart',
    'lemma:cont_mdiff_on_chart_symm',
    'lemma:cont_mdiff_id',
    'lemma:smooth_id',
    'lemma:cont_mdiff_on_id',
    'lemma:smooth_on_id',
    'lemma:cont_mdiff_at_id',
    'lemma:smooth_at_id',
    'lemma:cont_mdiff_within_at_id',
    'lemma:smooth_within_at_id',
    'lemma:cont_mdiff_const',
    'lemma:cont_mdiff_one',
    'lemma:smooth_const',
    'lemma:smooth_one',
    'lemma:cont_mdiff_on_const',
    'lemma:cont_mdiff_on_one',
    'lemma:smooth_on_const',
    'lemma:smooth_on_one',
    'lemma:cont_mdiff_at_const',
    'lemma:cont_mdiff_at_one',
    'lemma:smooth_at_const',
    'lemma:smooth_at_one',
    'lemma:cont_mdiff_within_at_const',
    'lemma:cont_mdiff_within_at_one',
    'lemma:smooth_within_at_const',
    'lemma:smooth_within_at_one',
    'lemma:cont_mdiff_of_support',
    'lemma:cont_mdiff_within_at_iff_cont_diff_within_at',
    'lemma:cont_mdiff_at_iff_cont_diff_at',
    'lemma:cont_mdiff_on_iff_cont_diff_on',
    'lemma:cont_mdiff_iff_cont_diff',
    'lemma:cont_mdiff_on.continuous_on_tangent_map_within_aux',
    'lemma:cont_mdiff_on.cont_mdiff_on_tangent_map_within_aux',
    'theorem:cont_mdiff_on.cont_mdiff_on_tangent_map_within',
    'theorem:cont_mdiff_on.continuous_on_tangent_map_within',
    'theorem:cont_mdiff.cont_mdiff_tangent_map',
    'theorem:cont_mdiff.continuous_tangent_map',
    'lemma:basic_smooth_vector_bundle_core.cont_mdiff_proj',
    'lemma:basic_smooth_vector_bundle_core.smooth_proj',
    'lemma:basic_smooth_vector_bundle_core.cont_mdiff_on_proj',
    'lemma:basic_smooth_vector_bundle_core.smooth_on_proj',
    'lemma:basic_smooth_vector_bundle_core.cont_mdiff_at_proj',
    'lemma:basic_smooth_vector_bundle_core.smooth_at_proj',
    'lemma:basic_smooth_vector_bundle_core.cont_mdiff_within_at_proj',
    'lemma:basic_smooth_vector_bundle_core.smooth_within_at_proj',
    'lemma:basic_smooth_vector_bundle_core.smooth_const_section',
    'lemma:tangent_bundle.cont_mdiff_proj',
    'lemma:tangent_bundle.smooth_proj',
    'lemma:tangent_bundle.cont_mdiff_on_proj',
    'lemma:tangent_bundle.smooth_on_proj',
    'lemma:tangent_bundle.cont_mdiff_at_proj',
    'lemma:tangent_bundle.smooth_at_proj',
    'lemma:tangent_bundle.cont_mdiff_within_at_proj',
    'lemma:tangent_bundle.smooth_within_at_proj',
    'def:tangent_bundle.zero_section',
    'lemma:tangent_bundle.smooth_zero_section',
    'lemma:tangent_bundle.tangent_map_tangent_bundle_pure',
    'lemma:cont_mdiff_within_at.prod_mk',
    'lemma:cont_mdiff_within_at.prod_mk_space',
    'lemma:cont_mdiff_at.prod_mk',
    'lemma:cont_mdiff_at.prod_mk_space',
    'lemma:cont_mdiff_on.prod_mk',
    'lemma:cont_mdiff_on.prod_mk_space',
    'lemma:cont_mdiff.prod_mk',
    'lemma:cont_mdiff.prod_mk_space',
    'lemma:smooth_within_at.prod_mk',
    'lemma:smooth_within_at.prod_mk_space',
    'lemma:smooth_at.prod_mk',
    'lemma:smooth_at.prod_mk_space',
    'lemma:smooth_on.prod_mk',
    'lemma:smooth_on.prod_mk_space',
    'lemma:smooth.prod_mk',
    'lemma:smooth.prod_mk_space',
    'lemma:cont_mdiff_within_at_fst',
    'lemma:cont_mdiff_at_fst',
    'lemma:cont_mdiff_on_fst',
    'lemma:cont_mdiff_fst',
    'lemma:smooth_within_at_fst',
    'lemma:smooth_at_fst',
    'lemma:smooth_on_fst',
    'lemma:smooth_fst',
    'lemma:cont_mdiff_within_at_snd',
    'lemma:cont_mdiff_at_snd',
    'lemma:cont_mdiff_on_snd',
    'lemma:cont_mdiff_snd',
    'lemma:smooth_within_at_snd',
    'lemma:smooth_at_snd',
    'lemma:smooth_on_snd',
    'lemma:smooth_snd',
    'lemma:smooth_iff_proj_smooth',
    "lemma:cont_mdiff_within_at.prod_map'",
    'lemma:cont_mdiff_within_at.prod_map',
    'lemma:cont_mdiff_at.prod_map',
    "lemma:cont_mdiff_at.prod_map'",
    'lemma:cont_mdiff_on.prod_map',
    'lemma:cont_mdiff.prod_map',
    'lemma:smooth_within_at.prod_map',
    'lemma:smooth_at.prod_map',
    'lemma:smooth_on.prod_map',
    'lemma:smooth.prod_map',
    'lemma:cont_mdiff_within_at_pi_space',
    'lemma:cont_mdiff_on_pi_space',
    'lemma:cont_mdiff_at_pi_space',
    'lemma:cont_mdiff_pi_space',
    'lemma:smooth_within_at_pi_space',
    'lemma:smooth_on_pi_space',
    'lemma:smooth_at_pi_space',
    'lemma:smooth_pi_space',
    'lemma:continuous_linear_map.cont_mdiff',
    'lemma:smooth_smul',
    'lemma:smooth.smul',
    'lemma:smooth_on.smul',
    'lemma:smooth_at.smul',
  ])
# ---
# name: test_parse_lean_ess_sup_fixture
  list([
    'def:ess_sup',
    'def:ess_inf',
    'lemma:ess_sup_congr_ae',
    'lemma:ess_inf_congr_ae',
    'lemma:ess_sup_eq_Inf',
    'lemma:ess_sup_measure_zero',
    'lemma:ess_inf_measure_zero',
    'lemma:ess_sup_mono_ae',
    'lemma:ess_inf_mono_ae',
    'lemma:ess_sup_const',
    'lemma:ess_sup_le_of_ae_le',
    'lemma:ess_inf_const',
    'lemma:le_ess_inf_of_ae_le',
    'lemma:ess_sup_const_bot',
    'lemma:ess_inf_const_top',
    'lemma:order_iso.ess_sup_apply',
    'lemma:order_iso.ess_inf_apply',
    'lemma:ess_sup_mono_measure',
    "lemma:ess_sup_mono_measure'",
    'lemma:ess_inf_antitone_measure',
    'lemma:ess_sup_smul_measure',
    'lemma:ess_sup_comp_le_ess_sup_map_measure',
    'lemma:measurable_embedding.ess_sup_map_measure',
    'lemma:ess_sup_map_measure_of_measurable',
    'lemma:ess_sup_map_measure',
    'lemma:ae_lt_of_ess_sup_lt',
    'lemma:ae_lt_of_lt_ess_inf',
    'lemma:ess_sup_indicator_eq_ess_sup_restrict',
    'lemma:ennreal.ae_le_ess_sup',
    'lemma:ennreal.ess_sup_eq_zero_iff',
    'lemma:ennreal.ess_sup_const_mul',
    'lemma:ennreal.ess_sup_add_le',
    'lemma:ennreal.ess_sup_liminf_le',
  ])
# ---
# name: test_parse_lean_fermat4_fixture
  list([
    'def:fermat_42',
    'lemma:fermat_42.comm',
    'lemma:fermat_42.mul',
    'lemma:fermat_42.ne_zero',
    'def:fermat_42.minimal',
    'lemma:fermat_42.exists_minimal',
    'lemma:fermat_42.coprime_of_minimal',
    'lemma:fermat_42.minimal_comm',
    'lemma:fermat_42.neg_of_minimal',
    'lemma:fermat_42.exists_odd_minimal',
    'lemma:fermat_42.exists_pos_odd_minimal',
    'lemma:int.coprime_of_sq_sum',
    "lemma:int.coprime_of_sq_sum'",
    'lemma:fermat_42.not_minimal',
    'lemma:not_fermat_42',
    'theorem:not_fermat_4',
  ])
# ---
# name: test_parse_lean_filtered_colimits_fixture
  list([
    'abbreviation:SemiRing.filtered_colimits.R',
    'def:SemiRing.filtered_colimits.colimit',
    'def:SemiRing.filtered_colimits.colimit_cocone',
    'def:SemiRing.filtered_colimits.colimit_cocone_is_colimit',
    'abbreviation:CommSemiRing.filtered_colimits.R',
    'def:CommSemiRing.filtered_colimits.colimit',
    'def:CommSemiRing.filtered_colimits.colimit_cocone',
    'def:CommSemiRing.filtered_colimits.colimit_cocone_is_colimit',
    'abbreviation:Ring.filtered_colimits.R',
    'def:Ring.filtered_colimits.colimit',
    'def:Ring.filtered_colimits.colimit_cocone',
    'def:Ring.filtered_colimits.colimit_cocone_is_colimit',
    'abbreviation:CommRing.filtered_colimits.R',
    'def:CommRing.filtered_colimits.colimit',
    'def:CommRing.filtered_colimits.colimit_cocone',
    'def:CommRing.filtered_colimits.colimit_cocone_is_colimit',
  ])
# ---
# name: test_parse_lean_to_additive_fixture
  list([
    'structure:to_additive.value_type',
  ])
# ---
